### PR TITLE
Add NVIDIA NGC image aliases

### PR DESCRIPTION
**Pull request by 0x2b3bfa0**
[NVIDIA NGC](https://docs.nvidia.com/ngc/ngc-deploy-public-cloud) offers machine and container images with **GPU drivers** and **Docker with GPU support** out of the box for [`aws`](https://docs.nvidia.com/ngc/ngc-deploy-public-cloud/ngc-aws/#ami-overview), [`az`](https://docs.nvidia.com/ngc/ngc-deploy-public-cloud/ngc-azure/#overview), [`gcp`](https://docs.nvidia.com/ngc/ngc-deploy-public-cloud/ngc-gcp/index.html#overview), [`k8s`](https://docs.nvidia.com/ngc/ngc-deploy-public-cloud/ngc-deploy-containers/index.html) and other platforms.

We are “maintaining” custom AWS machine images (https://github.com/iterative/terraform-provider-iterative/pull/206) that offer the same packages, plus Node.js for running the CML source distribution. After [cml#825](https://github.com/iterative/cml/issues/825) we don't need Node.js anymore, and it makes more sense to use NVIDIA images instead of publishing our own.

#### Related
* https://github.com/iterative/terraform-provider-iterative/issues/174
* https://github.com/iterative/dvc.org/pull/3246
* https://github.com/iterative/dvc.org/pull/3087

## Example

```hcl
resource "iterative_task" "example" {
  name  = "example"
  cloud = "aws"

  machine   = "m+v100"
  disk_size = "32"
  image     = "nvidia"

  script = <<-END
    #!/bin/sh
    nvidia-smi
    docker -v
  END
}
```

> 📖 _Note that the default disk size (30 GB) is not enough for most of these images; it should be explicitly increased._

#### `aws`

```console
$ aws ec2 describe-images --region=us-east-1 --owners=aws-marketplace --filter=Name=name,Values="NVIDIA*"
{
    "Images": [
        {
            "Architecture": "x86_64",
            "OwnerId": "679593333241",
            "Name": "NVIDIA Deep Learning  AMI v21.02.2-46a68101-e56b-41cd-8e32-631ac6e5d02b",
            "ProductCodes": [
                {
                    "ProductCodeId": "46kqvw7d9nbo0v2bfgi8z26gb",
                    "ProductCodeType": "marketplace"
                }
            ],
            ...
        },
        ...
    ]
}
```

```hcl
image = "ubuntu@679593333241:x86_64:NVIDIA Deep Learning  AMI v21.02.2-46a68101-e56b-41cd-8e32-631ac6e5d02b"
```

> 📖 _Visit https://aws.amazon.com/marketplace/pp?sku=46kqvw7d9nbo0v2bfgi8z26gb to accept terms and subscribe._

#### `az`

```console
$ az vm image list --publisher=nvidia --all
[
  {
    "offer": "ngc_base_image_version_b",
    "publisher": "nvidia",
    "sku": "gen2_21-11-0",
    "urn": "nvidia:ngc_base_image_version_b:gen2_21-11-0:21.11.0",
    "version": "21.11.0"
  },
  ...
]
```

```hcl
image = "ubuntu@nvidia:ngc_base_image_version_b:gen2_21-11-0:21.11.0#plan"
```

> 📖  _Run the following [`az`](https://docs.microsoft.com/en-us/cli/azure/?view=azure-cli-latest) command to [accept the image terms](https://docs.microsoft.com/en-us/cli/azure/vm/image/terms?view=azure-cli-latest):_
> 
> ```console
> az vm image terms accept --urn nngc_base_image_version_b:gen2_21-06-0:21.06.0
> ```

#### `gcp`

```console
$ gcloud compute images list --project=nvidia-ngc-public --filter=name~nvidia-gpu-cloud
NAME                                               PROJECT            FAMILY  DEPRECATED  STATUS
nvidia-gpu-cloud-image-20211105                    nvidia-ngc-public                      READY
...
```

```hcl
image = "ubuntu@nvidia-ngc-public/nvidia-gpu-cloud-image-20211105"
```

<blockquote>

⚠️ _Drivers don't work out of the box because they require a manual reboot to finish the installation._
<details><summary>Gory details...</summary>
When logging in with an interactive shell, the following message is displayed:
<pre>
NVIDIA GPU Cloud (NGC) is an optimized software environment that requires the
latest NVIDIA drivers to operate. If you do not download the NVIDIA drivers at
this time, your instance will shut down. Would you like to download the latest
NVIDIA drivers so NGC can finish installing? (Y/n)
</pre>

See https://gist.github.com/eliask/cd14372aabcb871aaaa661a87e2a2b2d for some workarounds.

</details>
</blockquote>

#### `k8s`

Example for the  [`nvidia/cuda:11.6.0-runtime-ubuntu20.04`](https://hub.docker.com/layers/nvidia/cuda/11.6.0-runtime-ubuntu20.04/images/sha256-70df05504934e38bec692f9f50365fdf66219d0f036619d1aa7c83dc0ee3024b?context=explore) tag:

```hcl
image = "nvidia/cuda:11.6.0-runtime-ubuntu20.04
```

See https://hub.docker.com/r/nvidia/cuda for a comprehensive list of tags.

### CODE REVIEW

File: task/az/resources/resource_virtual_machine_scale_set.go [hunk 1]
```diff
@@ -185,6 +187,14 @@ func (v *VirtualMachineScaleSet) Create(ctx context.Context) error {
 		},
 	}
 
+	if plan == "#plan" {
+		settings.Plan = &compute.Plan{
+			Publisher: to.StringPtr(publisher),
+			Product:   to.StringPtr(offer),
+			Name:      to.StringPtr(sku),
+		},
```
Comment (0x2b3bfa0) [inline task/az/resources/resource_virtual_machine_scale_set.go:195]:
```suggestion
		}
```

**Comment by DavidGOrtega** [comment]
> After https://github.com/iterative/cml/issues/825 we don't need Node.js anymore,

💯 

Should we extend this pr to machine and runner?

**Comment by 0x2b3bfa0** [comment]
> _Should we extend this pull request to machine and runner?_

If we did, users would have to perform an extra step (accept the terms) before using `cml runner`

**Comment by 0x2b3bfa0** [comment]
Maybe we can default to just–in–time provisioning of vanilla Ubuntu 20.04 images, and then recommend switching to the Nvidia ones as an optimization?

File: task/aws/resources/data_source_image.go [hunk 2]
```diff
@@ -34,6 +34,7 @@ func (i *Image) Read(ctx context.Context) error {
 	image := i.Identifier
 	images := map[string]string{
 		"ubuntu": "ubuntu@099720109477:x86_64:*ubuntu/images/hvm-ssd/ubuntu-focal-20.04*",
+		"nvidia": "ubuntu@679593333241:x86_64:NVIDIA Deep Learning  AMI v21.02.2-*",
```
Comment (DavidGOrtega) [inline task/aws/resources/data_source_image.go:37]: nothing relevant but you have an extra space without the OCD that you don't have ringing?

Comment (0x2b3bfa0) [reply task/aws/resources/data_source_image.go:37]: 💯 You have eagle eyes! It turns out that the image name **has** two spaces in it.

Comment (0x2b3bfa0) [reply task/aws/resources/data_source_image.go:37]: When I wrote this pull request, I took the effort of triple–checking it with `aws ec2 describe-images` to make sure it wasn't a typo.

Comment (0x2b3bfa0) [reply task/aws/resources/data_source_image.go:37]:
```bash
aws ec2 describe-images --region us-east-1 --filter "Name=name,Values=NVIDIA Deep Learning  AMI v21.02.2-*"
```

**Pull request merged**